rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala

```diff
@@ -496,39 +502,86 @@ class DebruijnInterpreter[M[_]: Sync: Parallel: _cost](
+          for {
+            v <- evalSingleExpr(p)
+            result <- v.exprInstance match {
+                       case GInt(hs) => Expr(GInt(-hs)).pure[M]
+                       case GBigInt(hs) =>
+                         for {
+                           r <- Sync[M].delay(-hs)
+                           _ <- charge[M](bigIntNegation(r))
+                         } yield Expr(GBigInt(r))
+                       case other => OperatorNotDefined("Negation", other.typ).raiseError[M, Expr]
+                     }
+          } yield result
+
         case EMultBody(EMult(p1, p2)) =>
           for {
-            v1 <- evalToLong(p1)
-            v2 <- evalToLong(p2)
-            _  <- charge[M](MULTIPLICATION_COST)
-          } yield GInt(v1 * v2)
+            v1 <- evalSingleExpr(p1)
+            v2 <- evalSingleExpr(p2)
+            result <- (v1.exprInstance, v2.exprInstance) match {
+                       case (GInt(lhs), GInt(rhs)) =>
+                         charge[M](MULTIPLICATION_COST).as(Expr(GInt(lhs * rhs)))
+                       case (GBigInt(lhs), GBigInt(rhs)) =>
+                         charge[M](bigIntMultiplication(lhs, rhs)).as(Expr(GBigInt(lhs * rhs)))
+                       case (_: GInt, other) =>
+                         OperatorExpectedError("*", "Int", other.typ).raiseError[M, Expr]
+                       case (_: GBigInt, other) =>
+                         OperatorExpectedError("*", "BigInt", other.typ).raiseError[M, Expr]
+                       case (other, _) => OperatorNotDefined("*", other.typ).raiseError[M, Expr]
+                     }
+          } yield result
 
         case EDivBody(EDiv(p1, p2)) =>
           for {
-            v1 <- evalToLong(p1)
-            v2 <- evalToLong(p2)
-            _  <- charge[M](DIVISION_COST)
-          } yield GInt(v1 / v2)
+            v1 <- evalSingleExpr(p1)
+            v2 <- evalSingleExpr(p2)
+            result <- (v1.exprInstance, v2.exprInstance) match {
+                       case (GInt(lhs), GInt(rhs)) =>
+                         charge[M](DIVISION_COST).as(Expr(GInt(lhs / rhs)))
+                       case (GBigInt(lhs), GBigInt(rhs)) =>
+                         charge[M](bigIntDivision(lhs, rhs)).as(Expr(GBigInt(lhs / rhs)))
+                       case (_: GInt, other) =>
+                         OperatorExpectedError("/", "Int", other.typ).raiseError[M, Expr]
+                       case (_: GBigInt, other) =>
+                         OperatorExpectedError("/", "BigInt", other.typ).raiseError[M, Expr]
+                       case (other, _) => OperatorNotDefined("/", other.typ).raiseError[M, Expr]
+                     }
+          } yield result
 
         case EModBody(EMod(p1, p2)) =>
           for {
-            v1 <- evalToLong(p1)
-            v2 <- evalToLong(p2)
-            _  <- charge[M](MODULO_COST)
-          } yield GInt(v1 % v2)
+            v1 <- evalSingleExpr(p1)
+            v2 <- evalSingleExpr(p2)
```

Comment:
Changes to negation, multiplication, division and module will slightly change behavior not related to BigInt.
E.g. `- {2 | 2}` or `{2 | 2} * 2` will now produce different error messages. Please add this into PR's description so it's easier to track.